File: ActionUI/Common/ActionUIInsertion.swift
// Common/ActionUIInsertion.swift
//
// Types and errors used by ActionUIModel.insertElement / insertRow / removeElement.
// Containers declare which containers accept runtime insertions via ActionUIViewConstruction.insertableContainers.

import Foundation

// Position to insert a new element (or row) within a container's existing array.
// `before(siblingID:)` and `after(siblingID:)` apply only to flat containers — for row containers,
// rows have no synthetic identity, so use `at(_:)`, `append`, or `prepend`.
public enum InsertPosition {
    case append
    case prepend
    case at(Int)
    case before(siblingID: Int)
    case after(siblingID: Int)
}

// Shape of a container container. Determines whether the container holds a flat array of elements
// (children, destinations, toolbar, commands, columns) or a 2-D array of element rows
// (Grid's `rows`).
public enum ContainerShape {
    case flat
    case rows
}

// Errors thrown by insertElement / insertRow / removeElement.
public enum InsertError: Error, CustomStringConvertible {
    case windowNotFound(String)
    case parentNotFound(parentID: Int)
    case notAContainer(type: String)
    case unknownContainer(container: String, vaildContainers: [String])
    case containerRequired(vaildContainers: [String])
    case wrongMethod(container: String, expectedShape: ContainerShape, message: String)
    case unsupportedPositionForRowContainer(container: String)
    case positionOutOfBounds(index: Int, count: Int)
    case siblingNotFound(siblingID: Int, container: String)
    case idConflict(ids: [Int])
    case invalidJSON(String)
    case missingType
    case rootRemovalForbidden(rootID: Int)
    case viewNotFound(viewID: Int)
    case viewIsRowMember(viewID: Int)

    public var description: String {
        switch self {
        case .windowNotFound(let uuid):
            return "No WindowModel for windowUUID: \(uuid)"
        case .parentNotFound(let id):
            return "No element found with parentID \(id)"
        case .notAContainer(let type):
            return "Element type '\(type)' does not accept insertions (no insertableContainers declared)"
        case .unknownContainer(let container, let valid):
            return "Unknown container '\(container)'. Valid containers: \(valid.joined(separator: ", "))"
        case .containerRequired(let valid):
            return "container is required when the container exposes multiple containers: \(valid.joined(separator: ", "))"
        case .wrongMethod(let container, let expectedShape, let message):
            return "Container '\(container)' has shape \(expectedShape); \(message)"
        case .unsupportedPositionForRowContainer(let container):
            return "Row container '\(container)' does not support before/after sibling positions; rows have no addressable identity"
        case .positionOutOfBounds(let i, let count):
            return "Position \(i) is out of bounds for container of size \(count)"
        case .siblingNotFound(let id, let container):
            return "No element with id \(id) found in container '\(container)'"
        case .idConflict(let ids):
            return "ID conflict — elements with these IDs are already registered: \(ids)"
        case .invalidJSON(let detail):
            return "Invalid JSON: \(detail)"
        case .missingType:
            return "Element dictionary missing 'type' key"
        case .rootRemovalForbidden(let id):
            return "Cannot remove root element (id \(id))"
        case .viewNotFound(let id):
            return "No element found with viewID \(id)"
        case .viewIsRowMember(let id):
            return "viewID \(id) is a Grid cell — removing it removes only the cell, not its row. Rows are not addressable by id."
        }
    }
}

File: ActionUIAndroid/library/src/main/java/com/abracode/actionui/Common/ActionUIInsertion.kt
package com.abracode.actionui.Common

/**
 * Types and errors for the runtime structural-mutation API
 * ([ActionUIModel.insertElement] / [ActionUIModel.insertRow] /
 * [ActionUIModel.removeElement]). The Android counterpart of Swift's
 * `Common/ActionUIInsertion.swift`.
 *
 * Containers declare which of their named subview containers accept runtime
 * insertions via [ActionUIViewConstruction.insertableContainers]; the model
 * resolves a request against that map ([ContainerShape] picks the method) and
 * mutates the parent [ViewModel.dynamicSubviews] override that the shared build
 * entry point merges back in. See [WindowModel] for the mutation engine.
 */

/**
 * Position to insert a new element (or row) within a container's existing array.
 * [Before] / [After] apply only to flat containers - for row containers, rows
 * have no synthetic identity, so use [At], [Append], or [Prepend].
 */
sealed class InsertPosition {
    /** Insert at the end of the container. */
    object Append : InsertPosition()

    /** Insert at the front of the container. */
    object Prepend : InsertPosition()

    /** Insert at an explicit index (0..count). */
    data class At(val index: Int) : InsertPosition()

    /** Insert immediately before the flat-container sibling with this id. */
    data class Before(val siblingID: Int) : InsertPosition()

    /** Insert immediately after the flat-container sibling with this id. */
    data class After(val siblingID: Int) : InsertPosition()
}

/**
 * Shape of an insertable container: a flat array of elements (`children`,
 * `destinations`) or a 2-D array of element rows (`Grid`'s `rows`). Determines
 * whether [ActionUIModel.insertElement] (flat) or [ActionUIModel.insertRow]
 * (rows) applies.
 */
enum class ContainerShape { FLAT, ROWS }

/**
 * Errors thrown by [ActionUIModel.insertElement] / [ActionUIModel.insertRow] /
 * [ActionUIModel.removeElement]. One [Exception] subclass per Apple
 * `InsertError` case, with the same human-readable messages (ASCII hyphens in
 * place of Apple's em dashes, per the project's ASCII-only source rule).
 */
sealed class InsertError(message: String) : Exception(message) {
    class WindowNotFound(uuid: String) :
        InsertError("No WindowModel for windowUUID: $uuid")

    class ParentNotFound(parentID: Int) :
        InsertError("No element found with parentID $parentID")

    class NotAContainer(type: String) :
        InsertError("Element type '$type' does not accept insertions (no insertableContainers declared)")

    class UnknownContainer(container: String, validContainers: List<String>) :
        InsertError("Unknown container '$container'. Valid containers: ${validContainers.joinToString(", ")}")

    class ContainerRequired(validContainers: List<String>) :
        InsertError("container is required when the element exposes multiple containers: ${validContainers.joinToString(", ")}")

    class WrongMethod(container: String, expectedShape: ContainerShape, detail: String) :
        InsertError("Container '$container' has shape $expectedShape; $detail")

    class UnsupportedPositionForRowContainer(container: String) :
        InsertError("Row container '$container' does not support before/after sibling positions; rows have no addressable identity")

    class PositionOutOfBounds(index: Int, count: Int) :
        InsertError("Position $index is out of bounds for container of size $count")

    class SiblingNotFound(siblingID: Int, container: String) :
        InsertError("No element with id $siblingID found in container '$container'")

    class IdConflict(ids: List<Int>) :
        InsertError("ID conflict - elements with these IDs are already registered: $ids")

    class InvalidJSON(detail: String) :
        InsertError("Invalid JSON: $detail")

    object MissingType :
        InsertError("Element dictionary missing 'type' key")

    class RootRemovalForbidden(rootID: Int) :
        InsertError("Cannot remove root element (id $rootID)")

    class ViewNotFound(viewID: Int) :
        InsertError("No element found with viewID $viewID")

    class ViewIsRowMember(viewID: Int) :
        InsertError("viewID $viewID is a Grid cell - removing it removes only the cell, not its row. Rows are not addressable by id.")
}
